dlt/destinations/sql_client.py

```diff
@@ -299,18 +321,42 @@ def df(self, chunk_size: int = None, **kwargs: Any) -> Optional[DataFrame]:
         May use native pandas/arrow reader if available. Depending on
         the native implementation chunk size may vary.
         """
+        return next(self.iter_df(chunk_size=chunk_size))
+
+    def arrow(self, chunk_size: int = None, **kwargs: Any) -> Optional[ArrowTable]:
+        """Fetches results as data frame in full or in specified chunks.
+
+        May use native pandas/arrow reader if available. Depending on
+        the native implementation chunk size may vary.
+        """
+        return next(self.iter_arrow(chunk_size=chunk_size))
+
+    def iter_fetchmany(self, chunk_size: int) -> Generator[List[Tuple[Any, ...]], Any, Any]:
+        while True:
+            if not (result := self.fetchmany(chunk_size)):
+                return
+            yield result
+
+    def iter_df(self, chunk_size: int) -> Generator[DataFrame, None, None]:
         from dlt.common.libs.pandas_sql import _wrap_result
 
         columns = self._get_columns()
-        if chunk_size is None:
-            return _wrap_result(self.native_cursor.fetchall(), columns, **kwargs)
-        else:
-            df = _wrap_result(self.native_cursor.fetchmany(chunk_size), columns, **kwargs)
-            # if no rows return None
-            if df.shape[0] == 0:
-                return None
-            else:
-                return df
+
+        # if no chunk size, fetch all
+        if not chunk_size:
+            yield _wrap_result(self.fetchall(), columns)
+            return
+
+        # otherwise iterate over results in batch size chunks
+        for result in self.iter_fetchmany(chunk_size=chunk_size):
```

Comment:
I'm not quite sure if this part is right. The tests pass but I am creating a dataframe with a pyarrow backend so the function below is more efficient, I am not sure how it is done in the rest of the codebase though. are we using numpy backends or pyarrow backends there?

Reply:
```
@staticmethod
    def _query_iterator(
        result,
        exit_stack: ExitStack,
        chunksize: int,
        columns,
        index_col=None,
        coerce_float: bool = True,
        parse_dates=None,
        dtype: DtypeArg | None = None,
        dtype_backend: DtypeBackend | Literal["numpy"] = "numpy",
    ):
        """Return generator through chunked result set"""
        has_read_data = False
        with exit_stack:
            while True:
                data = result.fetchmany(chunksize)
                if not data:
                    if not has_read_data:
                        yield _wrap_result(
                            [],
                            columns,
                            index_col=index_col,
                            coerce_float=coerce_float,
                            parse_dates=parse_dates,
                            dtype=dtype,
                            dtype_backend=dtype_backend,
                        )
                    break

                has_read_data = True
                yield _wrap_result(
                    data,
                    columns,
                    index_col=index_col,
                    coerce_float=coerce_float,
                    parse_dates=parse_dates,
                    dtype=dtype,
                    dtype_backend=dtype_backend,
                )
```

our biggest challenge is to provide `dtype` in case of tables

Reply:
I am not really sure what you mean with the code snippet above. I can't find it in the code.

Reply:
this is how pandas converts row tuples into panda frames. we do it OK in sql_database code so you can close it